### DISCUSSION TITLE
PocketBase (framework) auto-login

### DISCUSSION THREAD

**Discussion by vanvanni** [pocketbase/pocketbase#1737] [Q&A]
Hi,

I have a question regarding using PocketBase as a framework. 

Would love to create a custom route that allows you to auto-login into the web-ui via a token or something. Would this be possible?

**Comment by ganigeorgiev** [comment]
The Admin UI frontned cannot be customized or extended. I'm not sure that I understand your use case but the easiest way to allow Admin UI autologin functionality probably would be to update the LocalStorage `pb_admin_auth` key with a generated Admin model and token:

   - create a custom endpoint that will return an admin token based on some criteria, eg:
     
        ```go
        app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
            // returns admin auth token by some special secret value
            e.Router.AddRoute(echo.Route{
                Method: http.MethodPost,
                Path:   "/secret-login",
                Handler: func(c echo.Context) error {
                    // fetch an admin based on some criteria,
                    // eg. using some trusted! request (header, query, body) parameter
                    secretEmail := c.Request().Header.Get("X-Custom-Auth")
                    admin, err := app.Dao().FindAdminByEmail(secretEmail)

                    if err != nil {
                        apis.NewUnauthorizedError("Invalid or missing token", err)
                    }

                    authToken, tokenErr := tokens.NewAdminAuthToken(app, admin)
                    if tokenErr != nil {
                        return apis.NewBadRequestError("Failed to create auth token.", tokenErr)
                    }

                    return c.JSON(200, map[string]any{
                        "admin": admin,
                        "token": authToken,
                    })
                },
                Middlewares: []echo.MiddlewareFunc{
                    apis.ActivityLogger(app),
                },
            })

            return nil
        })
        ```

   - create a static html page on the same domain (could be placed inside `pb_public` or rendered with your go endpoint) that will call your custom endpoint and will update the LocalStorage `pb_admin_auth` key with the serialized json response and then redirect to the admin homepage (_having the token in the LocalStorage is what "sign-in" the admin_).

**Reply by vanvanni** [reply]
Sorry if I were unclear on my use case. But I will host a instance for person X, and want them to one-click login into there PB instance from my dashboard.  Not really familar with Go yet but Ill have a look into your solution. Looks like it would help me :)

**Reply by dschissler** [reply]
@joedvries I think that you might be able to read the Pocketbase source to see how authentication is done with a Auth record.  All `Records` have the capability to be upgrade to auth records with built-in functionality.  So `users` is just a generic auth collection.

You then might be able to call some custom endpoint in the Instance from the Dashboard server to obtain a auth key that you then return to client.  Maybe then they'd need to refresh the token in the client to get the auth user serialized into local storage?

For communicating from server to server you could use [charmbracelet/ssh](https://github.com/charmbracelet/ssh) to make your own SSH server that runs Cobra commands running along side the Pocketbase web server.  Then you just have to setup the public key.  It's really nice because it makes the SSH endpoint to not actually be a shell, and so it's much more controllable than normal SSH as a system user.

**Reply by vanvanni** [reply]
@dschissler That is quite interesting. Ill be able to read the code, writing is somewhat harder since I mainly write Node and PHP. But Ill figure it out I think. And Ill have a look at the package you linked, THANKS! Ill have a little investigation tonight.
